File: lab6/lab6-2.java
public class MinCostSearch {
    public static int minCostSearch(int[] k, int[] f) {
        int n = k.length;
        int[][] cost = new int[n][n];

        for (int i = 0; i < n; i++) {
            cost[i][i] = f[i];
        }

        for (int length = 2; length <= n; length++) {
            for (int i = 0; i <= n - length; i++) {
                int j = i + length - 1;
                cost[i][j] = Integer.MAX_VALUE;

                int totalFreq = 0;
                for (int x = i; x <= j; x++) {
                    totalFreq += f[x];
                }

                for (int r = i; r <= j; r++) {
                    int leftCost = (r > i) ? cost[i][r - 1] : 0;
                    int rightCost = (r < j) ? cost[r + 1][j] : 0;
                    cost[i][j] = Math.min(cost[i][j], leftCost + rightCost + totalFreq);
                }
            }
        }

        return cost[0][n - 1];
    }

    public static void main(String[] args) {
        // Example data
        int[] k = {5, 6};  
        int[] f = {17, 25};  

        int minCost = minCostSearch(k, f);
        System.out.println("Хамгийн бага өртөг: " + minCost);
    }
}

File: lab6/lab6-2.py
def min_cost_search(k, f):
    n = len(k)  
    cost = [[0] * n for _ in range(n)]  

    for i in range(n):
        cost[i][i] = f[i]  
        
    for length in range(2, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            cost[i][j] = float('inf')  
            total_freq = sum(f[i:j + 1])  
            for r in range(i, j + 1):
                left_cost = cost[i][r - 1] if r > i else 0
                right_cost = cost[r + 1][j] if r < j else 0
                cost[i][j] = min(cost[i][j], left_cost + right_cost + total_freq)

    return cost[0][n - 1]

# Жишээ
k = [5, 6]  # Түлхүүрүүд
f = [17, 25]  # Хайлтын тоо

min_cost = min_cost_search(k, f)
print("Хамгийн бага өртөг:", min_cost)
